### PR TITLE
Add git-metadata to exports for datadog-ci

### PR DESCRIPTION
**Pull request by IvanTopolcic**
### What and why?

This PR adds a few custom functions for uploading git-metadata and allowing use in a library instead of as a CLI. Serverless needs this functionality to use as part of other deployment-related libraries, such as datadog-cdk-constructs, datadog-serverless-plugin, etc.

### How?

This PR duplicates and simplifies some functionality. I didn't want to refactor the datadog-ci code because I did not want to break datadog-ci for our users, so I re-used some of the code and made changes to it so that it could be used as a library instead of CLI. The majority of this work was making parameters arguments to functions instead of getting them from the environment. We also want it to silently fail in the case that `git` was not installed (as `git` is not a requirement for our deployment tools). In that case, it would just not install the source code integration.

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)

### CODE REVIEW

File: src/commands/synthetics/crypto.ts [hunk 1]
```diff
@@ -1,5 +1,4 @@
 import {generateKeyPairSync} from 'crypto'
-
```
Comment (m-rousse) [inline src/commands/synthetics/crypto.ts:2]: Our convention is to separate imports from Node API, then 3rd party libraries, then project imports.
Could we keep this blank line? (same for other removed blank lines in imports of other files of this PR)

File: src/commands/git-metadata/git.ts [hunk 2]
```diff
@@ -25,6 +24,25 @@ export const newSimpleGit = async (): Promise<simpleGit.SimpleGit> => {
   return simpleGit.gitP(options)
 }
 
+// Returns a configured SimpleGit or undefined
+// in the case that the user does not have git installed
+export const newSimpleGitOrFail = async (): Promise<simpleGit.SimpleGit> => {
```
Comment (m-rousse) [inline src/commands/git-metadata/git.ts:29]: Could this function be factorized with `newSimpleGit` as it shares most of its implementation?

File: src/commands/git-metadata/git.ts [hunk 3]
```diff
@@ -93,3 +111,16 @@ export const getCommitInfo = async (
 
   return new CommitInfo(hash, remote, trackedFiles)
 }
+
+export const getCommitInfoBasic = async (git: simpleGit.SimpleGit): Promise<CommitInfo> => {
```
Comment (m-rousse) [inline src/commands/git-metadata/git.ts:115]: Could this function be factorized with `getCommitInfo`?
As the difference is whether the error is thrown of printed on the standard output, it could always be thrown and the display of the error to the stdout should be handled by the command itself.
The model we used in the synthetics command was to export as a library all functions that are "general" and implemented in `cli.ts` anything that is specific to the CLI usage/interaction of the command. With this model, handling the standard output of the function should be done in `cli.ts`.

File: yarn.lock [hunk 4]
```diff
@@ -2,35 +2,29 @@
 # yarn lockfile v1
```
Comment (m-rousse) [inline yarn.lock]: It seems the lockfile was almost completely rebuilt, though I do not see any `package.json` changes.
Are these changes necessary? If not, could you revert this file?

File: src/commands/git-metadata/library.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        },
+        retries: 5,
+      })
+
+      if (status !== UploadStatus.Success) {
+        throw new Error('Error uploading commit information.')
+      }
+
+      return payload.hash
+    } catch (error) {
```
Comment (m-rousse) [inline src/commands/git-metadata/library.ts:71]: This catch does not change the default behavior, could it be removed?

File: src/commands/git-metadata/__tests__/git.test.ts [hunk 6]
```diff
@@ -1,4 +1,5 @@
-import {getCommitInfo, gitRemote, stripCredentials} from '../git'
+import * as simpleGit from 'simple-git'
+import {getCommitInfo, gitRemote, newSimpleGitOrFail, stripCredentials} from '../git'
```
Comment (m-rousse) [inline src/commands/git-metadata/__tests__/git.test.ts:2]:
```suggestion

import {getCommitInfo, gitRemote, newSimpleGitOrFail, stripCredentials} from '../git'
```

File: src/commands/git-metadata/__tests__/library.test.ts [hunk 7]
```diff
@@ -0,0 +1,77 @@
+import {SimpleGit} from 'simple-git'
+import {SourceCodeIntegration} from '..'
```
Comment (m-rousse) [inline src/commands/git-metadata/__tests__/library.test.ts:2]:
```suggestion

import {SourceCodeIntegration} from '..'
```

File: src/commands/git-metadata/library.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+      isRepo = simpleGit.checkIsRepo()
+    } catch {
+      return false
+    }
+
+    // Only enable if the system has `git` installed and we're in a git repo
+    return apiKey !== undefined && isRepo
+  }
+
+  public async addSourceCodeIntegration(): Promise<string> {
```
Comment (jybp) [inline src/commands/git-metadata/library.ts:32]: Should this be renamed to something like `uploadGitMetadata` since `add` doesn't really describe what's happening under-the-hood? Just want to make sure since it'll be imported elsewhere.

File: src/commands/git-metadata/git.ts (hunk 3 above)
Comment (jybp) [reply src/commands/git-metadata/git.ts:115]: On the global model: In our case the cli specific logic is inside `upload.ts`. I agree that it'd be great to reduce code duplication if possible so that changes don't have to be applied twice in `upload.ts` and `library.ts`. As long as it's not a complete headache that is.

File: src/commands/synthetics/crypto.ts (hunk 1 above)
Comment (IvanTopolcic) [reply src/commands/synthetics/crypto.ts:2]: Ah missed that thanks, will do!

File: src/commands/git-metadata/git.ts (hunk 2 above)
Comment (IvanTopolcic) [reply src/commands/git-metadata/git.ts:29]: I'm not really sure how to refactor this function because the original logical swallows the exceptions. newSimpleGit() will return a GitClient even if git is not installed, and subsequent calls to the gitclient will throw an error. In upload.ts, this is handled by checking if the return value of getCommitInfo is undefined, and getCommitInfo returns undefined by wrapping everything in a try catch and just returning undefined if an exception is thrown. To me, it makes more sense to fail early. Do you have any recommendations here?

File: src/commands/git-metadata/library.ts (hunk 8 above)
Comment (IvanTopolcic) [reply src/commands/git-metadata/library.ts]: Good catch! I think `uploadGitCommitHash` is descriptive enough.

File: src/commands/git-metadata/git.ts (hunk 3 above)
Comment (IvanTopolcic) [reply src/commands/git-metadata/git.ts:115]: I didn't want to touch the existing implementation because I wasn't sure if there were any specific reasons the code was written the way it was. I agree that it makes the most sense to me to add the base functionality to `library` or `git` or whatever else is exported, and leave the pretty-format and printing to the CLI command (`upload`). This way, we can just handle the errors in the functions that are calling the library/git commands.

 I didn't want to introduce breaking behavior and re-writing the code to throw an error instead of printing and returning undefined. If you'd like, I can go ahead and try to refactor this a little so that we're printing the error inside the calling code instead of inside the library. @jybp and @m-rousse, thoughts?

File: src/commands/git-metadata/git.ts (hunk 2 above)
Comment (IvanTopolcic) [reply src/commands/git-metadata/git.ts]: After a discussion with @jybp I've refactored the Upload command to use the library.

File: src/commands/git-metadata/git.ts (hunk 3 above)
Comment (IvanTopolcic) [reply src/commands/git-metadata/git.ts]: I've gone ahead and refactored `getCommitInfo` to just throw the errors. The onus should be on the caller to catch and handle these appropriately.

File: src/commands/git-metadata/library.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  let isRepo
+  try {
+    simpleGit = await newSimpleGit()
+    isRepo = simpleGit.checkIsRepo()
+  } catch {
+    return false
+  }
+
+  // Only enable if the system has `git` installed and we're in a git repo
+  return apiKey !== undefined && isRepo
```
Comment (m-rousse) [inline src/commands/git-metadata/library.ts:19]: Nit: We try to avoid `let` as much as possible, you can use `return` in the `try` instead
```suggestion
  try {
    const simpleGit = await newSimpleGit()
    const isRepo = simpleGit.checkIsRepo()

    // Only enable if the system has `git` installed and we're in a git repo
    return apiKey !== undefined && isRepo
  } catch {
    return false
  }
```

**Review by m-rousse — APPROVED** [review]
This LGTM, thanks for bearing with the review! :+1:

(Also one CI check failed, once fixed this is OK in my opinion)

File: src/commands/git-metadata/library.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  try {
+    const simpleGit = await newSimpleGit()
+    const isRepo = simpleGit.checkIsRepo()
+
+    // Only enable if the system has `git` installed and we're in a git repo
+    return apiKey !== undefined && isRepo
+  } catch {
+    return false
+  }
+}
```
Comment (jybp) [inline src/commands/git-metadata/library.ts:18]: nit: Just wondering if having `apiKey !== undefined` is really necessary in this exported func since it's passed as arg. Shouldn't it be checked by whoever call this function?  
Without it the function could be renamed to something like `isCwdInGitRepository` making what is being done explicit.

**Review by jybp — APPROVED** [review]
Thanks it looks great! I've run the command with the PR's code and it's uploading repository data properly. It seems safe to merge 👍
